### PR TITLE
Add access policies integration

### PR DESCRIPTION
**Pull request by MarcosDY**
Consume allowed spiffe ids from access policies configured in EC.  
Based in retrieved access policies nginx will prevent access.

### CODE REVIEW

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp [hunk 1]
```diff
@@ -5,44 +5,48 @@ extern "C" {
...
+using grpc::StatusCode;
 
-bool firstFetch = true;
-void updatedCallback(X509SVIDResponse x509SVIDResponse);
-spiffe::WorkloadAPIClient workloadClient(updatedCallback);
+bool access_policies_updated = false;
+
+void svid_updated_callback(X509SVIDResponse x509SVIDResponse);
+void access_policy_updated_callback(emissary::api::proxy::AccessPolicyResponse accessPolicyResponse);
+void update_access_policies();
```
Comment (MarcosDY) [inline ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: never used

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp [hunk 2]
```diff
@@ -5,44 +5,48 @@ extern "C" {
...
+static std::vector<std::string> v_accepted_spiffe_ids;
+std::vector<std::string> get_accepted_spiffe_ids(emissary::api::proxy::AccessPolicyResponse accessPolicyResponse);
+spiffe::WorkloadAPIClient workloadClient(svid_updated_callback);
+scytale::emissary::ProxyAPIClient proxyClient(access_policy_updated_callback);
 static void log(std::string message, std::string arg);
 static void log(std::string message);
 static int xname_cmp(const X509_NAME * const *a, const X509_NAME * const *b);
                      
-extern "C" ngx_int_t create_spiffe_thread(ngx_ssl_t *ssl, ngx_flag_t is_client, ngx_int_t depth);
+extern "C" ngx_int_t create_spiffe_thread(ngx_ssl_t *ssl, ngx_flag_t is_client, ngx_int_t depth, ngx_array_t *spiffe_ids, ngx_pool_t  *pool);
```
Comment (MarcosDY) [inline ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: `, ngx_array_t *spiffe_ids, ngx_pool_t  *pool ` are no longer necessary

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp [hunk 3]
```diff
@@ -5,44 +5,48 @@ extern "C" {
...
 extern "C" ngx_int_t is_certificates_updated();
+extern "C" ngx_int_t is_spiffe_id_accepted(unsigned char *spiffe_id);
 ngx_int_t ngx_ssl_spiffe_reload_certificate(ngx_ssl_t *ngx_ssl, std::string cert_der, std::string key_der);
 ngx_int_t ngx_ssl_spiffe_reload_client_certificate(ngx_ssl_t *ngx_ssl, std::string bundle_der);
 ngx_int_t ngx_ssl_spiffe_reload_trusted_certificate(ngx_ssl_t *ngx_ssl, std::string bundle_der);
 ngx_int_t ngx_ssl_spiffe_add_all_ca(ngx_ssl_t *ngx_ssl, std::string bundle_der);
 
 typedef struct {
     ngx_ssl_t   **ngx_ssl;
+    ngx_array_t **accepted_spiffe_ids;
```
Comment (MarcosDY) [inline ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: we can remove it, no longer necessary

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp [hunk 4]
```diff
@@ -92,21 +118,64 @@ void updatedCallback(X509SVIDResponse x509SVIDResponse) {
...
+    }
+    proxyClient.SetSocketAddress("unix:" + socket_address);
+
+    Start:
+    proxyClient.FetchAccessPolicies();
+   
+    if (!proxyClient.GetFetchAccessPoliciesStatus().ok()) {
+        if (proxyClient.GetFetchAccessPoliciesStatus().error_code() == StatusCode::NOT_FOUND) {
+            // There is no access policy for this proxy. Clear the list of accepted SPIFFE IDs.
+            v_accepted_spiffe_ids = {};
```
Comment (MarcosDY) [inline ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: maybe we can add some log here just to notify people why we are no allowing any spiffe id

File: spiffe-support/src/http/modules/ngx_http_proxy_module.c [hunk 5]
```diff
@@ -4319,14 +4319,8 @@ ngx_http_proxy_set_ssl(ngx_conf_t *cf, ngx_http_proxy_loc_conf_t *plcf)
...
     if (plcf->upstream.ssl_spiffe) {
-        if(create_spiffe_thread(plcf->upstream.ssl, 0, plcf->ssl_verify_depth)
-            != NGX_OK)
-        {
-            return NGX_ERROR;
-        }
-
-        while (is_certificates_updated() != NGX_OK) {
-            // just wait
+        if (plcf->upstream.ssl_spiffe_accept == NULL) {
```
Comment (MarcosDY) [inline spiffe-support/src/http/modules/ngx_http_proxy_module.c]: no longer necessary, we are no using it

File: spiffe-support/src/http/modules/ngx_http_ssl_module.c [hunk 6]
```diff
@@ -781,15 +781,9 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
...
     //  
     if (conf->ssl_spiffe) {
-        if(create_spiffe_thread(&conf->ssl, 1, conf->verify_depth)
-            != NGX_OK)
-        {
-            return NGX_CONF_ERROR;
-        }
-        while (is_certificates_updated() != NGX_OK) {
-            // just wait
+        if (conf->ssl_spiffe_accept == NULL) {
```
Comment (MarcosDY) [inline spiffe-support/src/http/modules/ngx_http_ssl_module.c]: no longer necesary

File: spiffe-support/src/http/modules/ngx_http_ssl_module.h [hunk 7]
```diff
@@ -64,7 +64,7 @@ typedef struct {
 
 
 extern ngx_module_t  ngx_http_ssl_module;
-ngx_int_t create_spiffe_thread(ngx_ssl_t *ssl, ngx_flag_t is_clientt, ngx_int_t depth);
+ngx_int_t create_spiffe_thread(ngx_ssl_t *ssl, ngx_flag_t is_clientt, ngx_int_t depth, ngx_array_t *spiffe_ids, ngx_pool_t  *pool);
```
Comment (MarcosDY) [inline spiffe-support/src/http/modules/ngx_http_ssl_module.h]: `, ngx_array_t *spiffe_ids, ngx_pool_t  *pool,` no longer necesary

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp (hunk 4 above)
Comment (amartinezfayo) [reply ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: The Not Found error is already logged. Adding an additional log here would add two log entries for each failure. Maybe is not needed to log the fact that the list of SPIFFE IDs is cleared?

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp (hunk 2 above)
Comment (amartinezfayo) [reply ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: Good, I'll update the function definition.

File: ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp (hunk 3 above)
Comment (amartinezfayo) [reply ngx_http_fetch_spiffe_certs_module/ngx_http_fetch_spiffe_certs_module.cpp]: Good, I'll remote this.

File: spiffe-support/src/http/modules/ngx_http_proxy_module.c (hunk 5 above)
Comment (amartinezfayo) [reply spiffe-support/src/http/modules/ngx_http_proxy_module.c]: I'll remove it.

File: spiffe-support/src/http/modules/ngx_http_ssl_module.c (hunk 6 above)
Comment (amartinezfayo) [reply spiffe-support/src/http/modules/ngx_http_ssl_module.c]: I'll remove it.

File: spiffe-support/src/http/modules/ngx_http_ssl_module.h (hunk 7 above)
Comment (amartinezfayo) [reply spiffe-support/src/http/modules/ngx_http_ssl_module.h]: I'll remove this from the function definition.

**Pull request merged**